src/OpenColorIO/ops/Gamma/GammaOpUtils.cpp

```diff
@@ -0,0 +1,248 @@
+#include <math.h>
+
+#include <OpenColorIO/OpenColorIO.h>
+
+#include "BitDepthUtils.h"
+#include "ops/Gamma/GammaOpUtils.h"
+
+
+OCIO_NAMESPACE_ENTER
+{
+
+namespace
+{
+
+// Here we calculate the extra parameters used for the moncurve model.
+// The break point and slope of the linear segment are implied by the
+// gamma and offset.  The idea is that the linear segment has to meet
+// the power function at the point where the value and slope of the
+// two segments match.
+
+// The moncurve model would get a div by 0 error with gain=1, offset=0,
+// so the values need to get fudged slightly. We do that here rather
+// than during construction or validation so that the object may
+// contain the neat looking values since these are what would get written
+// to a ctf file.
+
+const double EPS = 1e-6;
+
+double monCurveGammaFwd(const GammaOpData::Params & p)
+{
+    return std::max(p[0], 1. + EPS);
+}
+
+double monCurveOffsetFwd(const GammaOpData::Params & p)
+{
+    const double offset = std::max(p[1], EPS);
+    return offset / (1. + offset);
+}
+
+double monCurveBreakFwd(const GammaOpData::Params & p)
+{
+    // Break point between the linear and power functions.
+    const double gamma = std::max(p[0], 1. + EPS);
+    const double offset = std::max(p[1], EPS);
+    return offset / (gamma - 1.);
+}
+
+double monCurveSlopeFwd(const GammaOpData::Params & p)
+{
+    // Slope of the linear segment.
+    const double gamma = std::max(p[0], 1. + EPS);
+    const double offset = std::max(p[1], EPS);
+    const double a = (gamma - 1.) / offset;
+    const double b = offset * gamma / (( gamma - 1.) * ( 1. + offset));
+    return a * pow(b, gamma);
+}
+
+double monCurveScaleFwd(const GammaOpData::Params & p)
+{
+    // This just rearranges the equation a little so we can get by
+    // with a single multiply rather than two.
+    const double offset = std::max(p[1], EPS);
+    return 1. / (1. + offset);
+}
+
+// Same stuff for the inverse of the forward function.
+
+double monCurveGammaRev(const GammaOpData::Params & p)
+{
+    return 1. / std::max(p[0], 1. + EPS);
+}
+
+double monCurveOffsetRev(const GammaOpData::Params & p)
+{
+    return std::max(p[1], EPS);
+}
+
+double monCurveBreakRev(const GammaOpData::Params & p)
+{
+    const double gamma = std::max(p[0], 1. + EPS);
+    const double offset = std::max(p[1], EPS);
+    const double a = offset * gamma;
+    const double b = (gamma - 1.) * (1. + offset);
+    return pow( a / b, gamma );
+}
+
+double monCurveSlopeRev(const GammaOpData::Params & p)
+{
+    const double gamma = std::max(p[0], 1. + EPS);
+    const double offset = std::max(p[1], EPS);
+    const double a = (gamma - 1.) / offset;
+    const double b = (1. + offset) / gamma;
+    return pow(a, gamma - 1.) * pow(b, gamma);
+}
+
+double monCurveScaleRev(const GammaOpData::Params & p)
+{
+    const double offset = std::max(p[1], EPS);
+    return 1. + offset;
+}
+
+
+
+};
+
+void ComputeParamsFwd(const GammaOpData::Params & gParams,

```

Comment:
Does this actually work?

Using the sRGB values provided elsewhere (`gamma=2.4, offset=0.055`) I don't get the correct breakpoint between the linear and power component of the curve. I get `0.03928571428571429` when calculated outside (in python numpy)...

That said the unittests should probably check against these known "truths" instead of always using arbitrary gamma and offset combinations.

Reply:
Sean, could you elaborate on why you think the break point is wrong?  For the sRGB case, the monCurveBreakFwd function above would return  0.055/(2.4-1) = 0.03928571428571429, exactly what you wrote.

Reply:
It should be 0.04045 no?

Reply:
Ah, im finding a descrepency based on what source you look at, fun.

So assuming the W3 site is correct then this should be fine.

Reply:
Is this discussion around the Rec.709/Rec.2020 gamma curve? When the industry moved to higher bit-depths they found that the constants were not high enough precision and there was a discontinuity at the break point. So they changed the constants to have another decimal place, which fixed that problem. The improved constants are fine when used with older content and lower bit-depths, so no need to keep both around; just use the improved ones.

Reply:
Dithermaster, we designed the code to avoid the problem you mention.  In various standards there are a lot of constants shown but in reality there are only two free parameters and the rest are calculated by someone and specified to various degrees of precision.  In our implementation the user specifies the two free parameters (the offset and exponent) and the other constants (break point, etc.) are calculated to machine precision so that the two curve segments will match precisely.

Reply:
> Dithermaster, we designed the code to avoid the problem you mention.
Very cool!